extensions/amp-story/0.1/pagination-buttons.js

```diff
@@ -0,0 +1,237 @@
+        (state, newState) =>
+          this.onButtonStateChange_(this.forwardButton_, state, newState));
+
+    /** @private @const {!AsyncValue<!ButtonStateDef>} */
+    this.backButtonState_ = AsyncValue.create(BackButtonStates.HIDDEN,
+        (state, newState) =>
+          this.onButtonStateChange_(this.backButton_, state, newState));
+  }
+
+  /**
+   * @param {!Document} doc
+   * @param {!function():Promise<boolean>} hasBookend
+   * @return {!PaginationButtons}
+   */
+  static create(doc, hasBookend) {
+    return new PaginationButtons(doc, hasBookend);
+  }
+
+  /**
+   * @param {!Element} el
+   * @param {!ButtonStateDef} state
+   * @param {!ButtonStateDef} newState
+   */
+  onButtonStateChange_(el, state, newState) {
+    el.classList.remove(state.className)
+    el.classList.add(newState.className);
+  }
+
+  /**
+   * @param {!Element} el
+   * @param {!Event} e
+   * @param {!AsyncValue<!ButtonStateDef>} state
+   */
+  maybeDispatchOnClick_(el, state) {
+    el.addEventListener('click', e => {
+      const eventType = state.get().triggers;
+      if (!eventType) {
+        return;
+      }
+      e.preventDefault();
+      dispatch(el, dev().assert(eventType), /* opt_bubbles */ true);
+    });
+  }
+
+  /** @param {!Element} element */
+  attach(element) {
+    setClassOnHover(this.forwardButton_, element, 'i-amphtml-story-next-hover');
+    setClassOnHover(this.backButton_, element, 'i-amphtml-story-prev-hover');
+
+    maybeDispatchOnClick(this.forwardButton_, this.forwardButtonState_);
+    maybeDispatchOnClick(this.backButton_, this.backButtonState_);
+
+    this.forwardButton_.classList.add(this.forwardButtonState_.get().className);
+    this.backButton_.classList.add(this.backButtonState_.get().className);
+
+    element.appendChild(this.forwardButton_);
+    element.appendChild(this.backButton_);
+  }
+
+  /** @param {!./navigation-state.StateChangeEventDef} event */
+  onNavigationStateChange(event) {
+    switch (event.type) {
+      case StateChangeType.ACTIVE_PAGE:
+        const {pageIndex, totalPages} = event.value;
+        if (pageIndex === totalPages - 1) {
+          this.onLastPageActive_();
+          return;
+        }
+        this.backButtonState_.set(
+            pageIndex === 0 ?
+              BackButtonStates.HIDDEN :
+              BackButtonStates.PREVIOUS_PAGE);
+        this.forwardButtonState_.set(ForwardButtonStates.NEXT_PAGE);
+        break;
+
+      case StateChangeType.BOOKEND_ENTER:
+        this.backButtonState_.set(BackButtonStates.CLOSE_BOOKEND);
+        this.forwardButtonState_.set(ForwardButtonStates.REPLAY);
+        break;
+
+      case StateChangeType.BOOKEND_EXIT:
+        this.onLastPageActive_();
+        break;
+    }
+  }
+
+  /** @private */
+  onLastPageActive_() {
+    this.backButtonState_.set(BackButtonStates.PREVIOUS_PAGE);
+    this.forwardButtonState_.set(this.hasBookend_().then(hasBookend =>
```

Comment:
Does `AsyncState` exist basically for this case? Interesting idea but I don't think the readability cost is worth it.

Reply:
Alright, simplified a bit.